package/src/netmon.h

```diff
@@ -0,0 +1,75 @@
+// Copyright (C) CERN, 2018
+//
+// Network monitoring class
+//
+// This is a more efficient network monitor that holds
+// its istreams in a structure, so that they don't have
+// to be opened and closed on each monitoring cycle.
+
+#ifndef PRMON_NETMON_H
+#define PRMON_NETMON_H 1
+
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+class netmon {
+ private:
+  // Which network interface paramters to measure
+  const std::vector<std::string> interface_params;
+
+  // Which network interfaces to monitor
+  std::vector<std::string> monitored_netdevs;
+
+  // Nested dictionary of network_if_streams[PARMETER][DEVICE][ISTREAM*]
+  std::unordered_map<
+      std::string,
+      std::unordered_map<std::string, std::unique_ptr<std::ifstream>>>
+      network_if_streams;
+
+  // Container for starting value stats, that are subtracted from measured
+  // values
+  std::unordered_map<std::string, unsigned long long> network_stats_start;
+
+  // Find all network interfaces on the system
+  std::vector<std::string> const get_all_network_devs();
+
+  // Open streams for each of the network interfaces and parameters
+  void open_interface_streams();
+
+  // Helper to map a device and parameter to a /sys filename
+  inline std::string const get_sys_filename(std::string device,
+                                            std::string param) {
+    std::string filename = "/sys/class/net/" + device + "/statistics/" + param;
+    return filename;
+  }
+
+  // Internal methods to read "raw" network stats
+  std::unordered_map<std::string, unsigned long long> read_raw_network_stats();
+  void read_raw_network_stats(
+      std::unordered_map<std::string, unsigned long long>& values);
+
+ public:
+  netmon(std::vector<std::string> netdevs);
+  netmon() : netmon(std::vector<std::string>{}){};
+
+  const std::vector<std::string> get_interface_paramter_names() {
+    return interface_params;
+  }
+
+  const std::vector<std::string> get_interface_names() {
+    return monitored_netdevs;
+  }
+
+  // Return network stats in a variable
+  std::unordered_map<std::string, unsigned long long> read_network_stats();
```

Comment:
Do we really need this overloaded definition of `read_raw_network_stats `? I don't think we use this anywhere. If we want to have this as means to `read`-only, shall it be a `get` method that returns a `const` instead? Maybe I'm missing something?

Reply:
But if you return `const` then it can't then be passed as a modifiable reference to be updated by `read_network_stats(&value)`, which is what's done later.

OTOH, I start to not like too much the overload either, especially as we have the public methods as well as the private "raw" versions. The client could just create the container itself and pass it in to the single updater method.

Reply:
Definitely, `const`ing them will break passing things around by reference for updating. I thought we might mistakenly update something and trying to see if there is a way to safeguard against that. Also I always find avoiding overloaded methods ease following the code but maybe that's personal :)